[WebGL] Log KHR_debug messages to system console

djg opened the pull request:
#### 8e4c281fd16129e538ce75fa7c6bf0f427ed882d
<pre>
[WebGL] Log KHR_debug messages to system console
<a href="https://bugs.webkit.org/show_bug.cgi?id=274197">https://bugs.webkit.org/show_bug.cgi?id=274197</a>
<a href="https://rdar.apple.com/128107146">rdar://128107146</a>

Reviewed by Kimmo Kinnunen and Mike Wyrzykowski.

To aid debugging WebGL apps, surface debug messages generated by the KHR_debug
extension for the GL_DEBUG_SOURCE_API message source to the system
console. Errors are formatted in the same manner as other errors resulting from
WebGL operations, where as all other message types are formatted as informational.

* LayoutTests/fast/canvas/webgl/debug-messages-to-console-expected.txt: Added.
* LayoutTests/fast/canvas/webgl/debug-messages-to-console.html: Added.
* LayoutTests/fast/canvas/webgl/index-validation-with-subsequent-draws.html:
* LayoutTests/fast/canvas/webgl/invalid-UTF-16.html:
* LayoutTests/fast/canvas/webgl/largeBuffer.html:
* LayoutTests/fast/canvas/webgl/webgl2-texStorage.html:
* LayoutTests/webgl/webgl-allow-shared.html:
* LayoutTests/webgl/webgl-multi-draw-noop.html:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::glEnumToErrorCode):
(WebCore::resolveGraphicsContextGLAttributes):
(WebCore::debugMessageTypeToString):
(WebCore::debugMessageSeverityToString):
(WebCore::WebGLRenderingContextBase::addDebugMessage):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::initialize):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::addDebugMessage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::wasLost):
(WebKit::RemoteGraphicsContextGLProxy::addDebugMessage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:

Canonical link: <a href="https://commits.webkit.org/279786@main">https://commits.webkit.org/279786@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d115a6cc047fa1a3a65de1f757b91cfddb57e36

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54465 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57744 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5196 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56768 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41412 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5186 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44102 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3484 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56560 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32015 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47170 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25238 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28813 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4499 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3339 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4713 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59335 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29691 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4873 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51527 "Found 1 new test failure: imported/w3c/web-platform-tests/server-timing/server_timing_header-parsing.https.html (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30844 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47259 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50897 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31831 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8070 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30636 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple reviewed (changes requested):
I think it's overly complex to add prefs for what it does.
Could we instead add code to route the messages to web inspector and remove all settings related code?

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -5627,6 +5628,77 @@ void WebGLRenderingContextBase::forceContextLost()
+    }
+}
+
+static ASCIILiteral debugIdToString(GCGLenum id)
```
this already exits as errorCodeToString

kkinnunen-apple replied:
E.g. the spec is that when source is api, type is error, id is of type gl error enum. There isn't "debug id to string" function per se, as that is a mapping of source, type, id -> string

djg replied:
> this already exits as errorCodeToString

But `GCGLErrorCode` is an bitfield not a gl error enum, so we still need `glErrorToErrorCode`.

I chose `debugIdToString` to keep with the pattern of the similar functions.

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -5627,6 +5628,77 @@ void WebGLRenderingContextBase::forceContextLost()
+        return;
+
+    if (type == GraphicsContextGL::DEBUG_TYPE_ERROR)
+        printToConsole(MessageLevel::Error, makeString("WebGL: "_s, debugIdToString(id), ": "_s, message));
```
MessageLevel::Error will produce a JS stack shot which we cannot do, as this is called within the IPC message dispatch.
You need to inline enough stuff from printToConsole to somewhat match the format that it's natural for the reader.

kkinnunen-apple reviewed (changes requested):
This needs tests to ensure that the feature works and keeps working.
E.g. write a js-test that calls few cases that output errors. the console messages should arrive to the expected.txt file when the test settings are correct. See other webgl tests how they turn on the console messages to output

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -407,6 +427,7 @@ static GraphicsContextGLAttributes resolveGraphicsContextGLAttributes(const WebG
     glAttributes.alpha = attributes.alpha;
     glAttributes.depth = attributes.depth;
     glAttributes.stencil = attributes.stencil;
+    glAttributes.stencil = attributes.stencil;
```
remove this

kkinnunen-apple commented on LayoutTests/fast/canvas/webgl/debug-messages-to-console.html:
```diff
@@ -0,0 +1,36 @@
+<!DOCTYPE html>
+<html>
+<head>
+<script src="/js-test-resources/js-test.js"></script>
```
I don't think this file exists, you should see an error in Web Inspector.

instead, you should 
use the .../../resources/js-test.js 

use
`jsTestIsAsync = true`

This will make the test wait for function call jsTestFinished()

kkinnunen-apple commented on LayoutTests/fast/canvas/webgl/debug-messages-to-console.html:
```diff
@@ -0,0 +1,36 @@
+gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA8, 32, 32, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
+  
+// CONSOLE MESSAGE: WebGL: INVALID_ENUM: Invalid enum provided.
+gl.drawElements(0, 0, 0, 0);
```
This would mean that the test runner will finish the test when the onload has fired.
This has the shortcoming that we don't know if the messages were delivered at that time.

Instead, you should / could: 
- force gl to be evaluated by calling gl.getErrror() as the last call (while (gl.getError() != 0); )
- set, say, 50ms timer and from timer callback call jsTestFinished();

djg committed:
[WebGL] Log KHR_debug messages to system console
https://bugs.webkit.org/show_bug.cgi?id=274197
rdar://128107146

Reviewed by Kimmo Kinnunen and Mike Wyrzykowski.

To aid debugging WebGL apps, surface debug messages generated by the KHR_debug
extension for the GL_DEBUG_SOURCE_API message source to the system
console. Errors are formatted in the same manner as other errors resulting from
WebGL operations, where as all other message types are formatted as informational.

* LayoutTests/fast/canvas/webgl/debug-messages-to-console-expected.txt: Added.
* LayoutTests/fast/canvas/webgl/debug-messages-to-console.html: Added.
* LayoutTests/fast/canvas/webgl/index-validation-with-subsequent-draws.html:
* LayoutTests/fast/canvas/webgl/invalid-UTF-16.html:
* LayoutTests/fast/canvas/webgl/largeBuffer.html:
* LayoutTests/fast/canvas/webgl/webgl2-texStorage.html:
* LayoutTests/webgl/webgl-allow-shared.html:
* LayoutTests/webgl/webgl-multi-draw-noop.html:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::glEnumToErrorCode):
(WebCore::resolveGraphicsContextGLAttributes):
(WebCore::debugMessageTypeToString):
(WebCore::debugMessageSeverityToString):
(WebCore::WebGLRenderingContextBase::addDebugMessage):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::initialize):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::addDebugMessage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::wasLost):
(WebKit::RemoteGraphicsContextGLProxy::addDebugMessage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:

Canonical link: https://commits.webkit.org/279786@main
